subsys/net/lib/sockets/socketpair.c

```diff
@@ -0,0 +1,1121 @@
+	if (spair->remote != -1) {
+		remote = z_get_fd_obj(spair->remote,
+			(const struct fd_op_vtable *)&spair_fd_op_vtable, 0);
+
+		if (remote != NULL) {
+			res = k_sem_take(&remote->sem, K_FOREVER);
+			if (res == 0) {
+				have_remote_sem = true;
+				remote->remote = -1;
+				res = k_poll_signal_raise(&remote->write_signal,
+					SPAIR_SIG_CANCEL);
+				__ASSERT(res == 0,
+					"k_poll_signal_raise() failed: %d",
+					res);
+			}
+		}
+	}
+
+	spair->remote = -1;
+
+	res = k_poll_signal_raise(&spair->read_signal, SPAIR_SIG_CANCEL);
+	__ASSERT(res == 0, "k_poll_signal_raise() failed: %d", res);
+
+	/* ensure no private information is released to the memory pool */
+	memset(spair, 0, sizeof(*spair));
+
+	k_free(spair);
+
+	if (remote != NULL && have_remote_sem) {
+		k_sem_give(&remote->sem);
+	}
+}
+
+/**
+ * Create a @ref spair (1/2 of a socketpair)
+ *
+ * The idea is to call this twice, but store the "local" side in the
+ * @ref spair.remote field initially.
+ *
+ * If both allocations are successful, then swap the @ref spair.remote
+ * fields in the two @ref spair instances.
+ */
+static struct spair *spair_new(void)
+{
+	struct spair *spair;
+
+	spair = k_malloc(sizeof(*spair));
+	if (spair == NULL) {
+		errno = ENOMEM;
+		goto out;
+	}
+	memset(spair, 0, sizeof(*spair));
+
+	/* initialize any non-zero default values */
+	spair->remote = -1;
+	spair->flags = SPAIR_FLAGS_DEFAULT;
+
+	k_sem_init(&spair->sem, 1, 1);
+	k_pipe_init(&spair->recv_q, spair->buf, sizeof(spair->buf));
+	k_poll_signal_init(&spair->write_signal);
+	k_poll_signal_init(&spair->read_signal);
+
+	spair->remote = z_reserve_fd();
+	if (spair->remote == -1) {
+		errno = ENFILE;
+		goto cleanup;
+	}
+
+	z_finalize_fd(spair->remote, spair,
+		      (const struct fd_op_vtable *)&spair_fd_op_vtable);
+
+	goto out;
+
+cleanup:
+	spair_delete(spair);
+	spair = NULL;
+
+out:
+	return spair;
+}
+
+int z_impl_zsock_socketpair(int family, int type, int proto, int *sv)
+{
+	int res;
+	size_t i;
+	struct spair *obj[2] = {};
+
+	if (family != AF_UNIX) {
+		errno = EAFNOSUPPORT;
+		res = -1;
+		goto errout;
+	}
+
+	if (type != SOCK_STREAM) {
+		errno = EPROTOTYPE;
+		res = -1;
+		goto errout;
+	}
+
+	if (proto != 0) {
+		errno = EPROTONOSUPPORT;
+		res = -1;
+		goto errout;
+	}
+
+	if (sv == NULL) {
+		/* not listed in the normative spec, but mimics Linux behaviour */
+		errno = EFAULT;
+		res = -1;
+		goto errout;
+	}
+
+	for (i = 0; i < 2; ++i) {
+		obj[i] = spair_new();
+		if (!obj[i]) {
+			res = -1;
+			goto cleanup;
+		}
+	}
+
+	/* connect the two endpoints */
+	swap32(&obj[0]->remote, &obj[1]->remote);
+
+	for (i = 0; i < 2; ++i) {
+		sv[i] = obj[i]->remote;
+		k_sem_give(&obj[0]->sem);
+	}
+
+	return 0;
+
+cleanup:
+	for (i = 0; i < 2; ++i) {
+		spair_delete(obj[i]);
+	}
+
+errout:
+	return res;
+}
+
+#ifdef CONFIG_USERSPACE
+int z_vrfy_zsock_socketpair(int family, int type, int proto,
+					  int *sv)
+{
+	int ret;
+	int tmp[2];
+
+	if (!Z_SYSCALL_MEMORY_WRITE(sv, sizeof(tmp))) {
```

Comment:
@pfalcon - removing the check for NULL in `z_impl_zsock_socketpair()` broke tests where `sv` was `NULL`. Not 100% sure why, but I guess it was a non-userspace test? I've opted to be extra paranoid and leave the check in `z_impl_zsock_socketpair()` and to put a couple of extra checks here.

Reply:
I'm not sure why you're asking me, the only concern I had re: this (or similar) line was a missing parens ;-). Sounds good to me anyway.